desc/particles.py

```diff
@@ -0,0 +1,1239 @@
+        sqrtg = jnp.linalg.norm(data["x_s"], axis=-1) * data["ds"]
+        self._chosen_idxs = jax.random.choice(
+            key=jax.random.PRNGKey(self.seed),
+            a=sqrtg.shape[0],
+            shape=(self.N,),
+            replace=True,
+            p=sqrtg / sqrtg.sum(),
+        )
+
+        # positions of the selected nodes in R, phi, Z coordinates
+        x = jnp.take(data["x"], self._chosen_idxs, axis=0)
+        params = field.params_dict
+
+        if model.frame == "flux":
+            if not isinstance(field, Equilibrium):
+                raise ValueError(
+                    "Mapping from lab to flux coordinates requires an Equilibrium. "
+                    "Please use Equilibrium object with the model."
+                )
+            zeta = jnp.take(data["phi"], self._chosen_idxs, axis=0)
+            # this is not the best guess, but most likely scenario for this class
+            # is to initialize particles on the magnetic axis, for sake of jitable
+            # implementation, we use rho=0, theta=0 as the guess
+            x_guess = jnp.array([jnp.zeros(self.N), jnp.zeros(self.N), zeta]).T
+            if not self.is_curve_magnetic_axis:
+                tol = 1e-8
+                x, out = field.map_coordinates(
+                    coords=x,
+                    inbasis=("R", "phi", "Z"),
+                    outbasis=("rho", "theta", "zeta"),
+                    maxiter=200,
+                    guess=x_guess,
+                    tol=tol,
+                    full_output=True,
+                )
+                x = eqx.error_if(
+                    x,
+                    (out[0] > tol).any(),
+                    "Mapping from lab to flux coordinates failed to achieve tolerance "
+                    f"{tol:.2e}. Make sure the curve lies in the equilibrium.",
+                )
+            else:
+                x = x_guess
+            if field.iota is None:
+                iota = field.get_profile("iota")
+                params["i_l"] = iota.params
+                kwargs["iota"] = iota
+        elif model.frame == "lab":
+            if isinstance(field, Equilibrium):
+                raise NotImplementedError(
+                    "If you have an Equilibrium object, you should use the model "
+                    "in flux frame. Since trying to integrate in lab frame will "
+                    "require multiple coordinate mapping, it is not implemented."
+                )
+
+        v = jnp.sqrt(2 * self.E / self.m)
+        vpar = jax.random.uniform(
+            key=jax.random.PRNGKey(self.seed),
+            shape=(v.size,),
+            minval=self.xi_min * v,
+            maxval=self.xi_max * v,
+        )
+
+        return super()._return_particles(
+            x=x, v=v, vpar=vpar, model=model, field=field, params=params, **kwargs
+        )
+
+
+class SurfaceParticleInitializer(AbstractParticleInitializer):
+    """Randomly sample particles starting on a surface.
+
```

Comment:
explain the sampling more, it is specifically sampling the positions randomly from the surface on the given discrete grid, with higher probability given to points with larger surface area. Then each particle additioally has its xi assigned randomly